sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -39,137 +64,178 @@ public sealed class ResourceIdentifier : IEquatable<ResourceIdentifier>, ICompar
 #pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
         public ResourceIdentifier(string resourceId)
         {
-            ResourceIdentifier id = Create(resourceId);
-            Init(id.Parent, id.ResourceType, id.Name, id.IsProviderResource);
-            _stringValue = resourceId;
-        }
+            Argument.AssertNotNullOrEmpty(resourceId, nameof(resourceId));
 
-        private ResourceIdentifier(ResourceIdentifier? parent, ResourceType resourceType, string resourceName, bool idProviderResource)
-        {
-            Init(parent ?? Root, resourceType, resourceName, idProviderResource);
-        }
-#pragma warning restore CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+            if (resourceId.Length == 1 && resourceId[0] == Separator)
+            {
+                //this is the same as Root but can't return that since this is a ctor
+                Init(null, ResourceType.Tenant, string.Empty, false, null);
+            }
+            else
+            {
+                ReadOnlySpan<char> remaining = resourceId.AsSpan();
 
-        private ResourceIdentifier(ResourceIdentifier parent, string resourceTypeName, string resourceName)
-            : this(parent, ChooseResourceType(resourceTypeName, parent), resourceName, false)
-        {
+                if (!remaining.StartsWith(SubscriptionStart.AsSpan()) && !remaining.StartsWith(ProviderStart.AsSpan()))
```

Comment:
I am not sure if the `AsSpan()` methods here make a difference. I believe the `StartsWith` method is already optimized for strings. @KrzysztofCwalina thoughts?

Reply:
Since `remaining` is a `ReadOnlySpan<char>` the SubscriptionStart has to be converted to AsSpan as well

Reply:
My point was that using the `Span<T>` APIs this way might not make a difference. The `StartsWith` API in the BCL is already optimized to the point where converting this to `Span<T>` might actually be slower.

Reply:
The asspan startswith is faster

|           Method |      Mean |     Error |    StdDev | Allocated |
|----------------- |----------:|----------:|----------:|----------:|
| StartsWithString | 41.937 ns | 0.0642 ns | 0.0536 ns |         - |
|   StartsWithSpan |  3.409 ns | 0.0131 ns | 0.0116 ns |         - |

```
    [Benchmark]
    public bool StartsWithString() => "/subscriptions/0c2f6471-1bf0-4dda-aec3-cb9272f09575/resourceGroups/myRg/providers/Microsoft.Compute/virtualMachines/myVm/subnets/mysubnet".StartsWith("/subscriptions/");
    [Benchmark]
    public bool StartsWithSpan() => "/subscriptions/0c2f6471-1bf0-4dda-aec3-cb9272f09575/resourceGroups/myRg/providers/Microsoft.Compute/virtualMachines/myVm/subnets/mysubnet".AsSpan().StartsWith("/subscriptions/".AsSpan());
```